t/unit/utils/test_functional.py

```diff
@@ -150,6 +152,56 @@ def build_generator():
     def test_nonzero__empty_iter(self):
         assert not regen(iter([]))
 
+    def test_deque(self):
+        original_list = [42]
+        d = collections.deque(original_list)
+        # Confirm that concretising a `regen()` instance repeatedly for an
+        # equality check always returns the original list
+        g = regen(d)
+        assert g == original_list
+        assert g == original_list
+
+    def test_repr(self):
+        def die():
+            raise AssertionError("Generator died")
+            yield None
+
+        # Confirm that `regen()` instances are not concretised when represented
+        g = regen(die())
+        assert "..." in repr(g)
+
+    @pytest.mark.xfail(raises=StopIteration, reason="XXX NEEDS AN ISSUE")
```

Comment:
This test is for an edge case where a repeatable iterator could raise an exception at some point before completion which would result in `regen()` stashing the elements it saw but never marking itself as done. Then the next time we yield elements from an iterator derived from the `regen()` instance, we'd yield the stashed elements, then repeat them from the underlying iterator if it's capable of giving them to us again.

IDK if this is something we can even bother handling. If the underlying iterator explodes, we could choose to mark the regen instance as `__done` and just give up on attempting to get more elements? If we want to go down that route, I can roll a change into this PR since it'd be pretty simple and still thematically relevant. If we think we need to do something more involved or optimistic, I'll make a follow up issue for discussion.

Reply:
I've made #6794 to track the discussion on this point. Could just be a thing we need to document, happy to have this PR merged without addressing that issue yet.